pyro/infer/csis.py

```diff
@@ -0,0 +1,161 @@
+from pyro.util import check_model_guide_match, warn_if_nan
+
+
+class CSIS(Importance):
+    """
+    Compiled Sequential Importance Sampling, allowing compilation of a guide
+    program to minimise KL(model posterior || guide), and inference with
+    importance sampling.
+
+    **Reference**
+    "Inference Compilation and Universal Probabilistic Programming" `pdf https://arxiv.org/pdf/1610.09900.pdf`
+
+    :param model: probabilistic model defined as a function. Must accept
+        keyword arguments with names of observed sites in model.
+    :param guide: guide function which is used as an approximate posterior. Must
+        accept keyword arguments with names of observed sites in model.
+    :param optim: a PyTorch optimizer
+    :type optim: torch.optim.Optimizer
+    :param num_inference_samples: The number of importance-weighted samples to
+        draw during inference.
+    :param training_batch_size: Number of samples to use to approximate the loss
+        before each gradient descent step during training.
+    """
+    def __init__(self,
+                 model,
+                 guide,
+                 optim,
+                 num_inference_samples=10,
+                 training_batch_size=10):
+        super(CSIS, self).__init__(model, guide, num_inference_samples)
+        self.model = model
+        self.guide = guide
+        self.optim = optim
+        self.training_batch_size = training_batch_size
+        self.validation_batch = None
+        self.args = []
+        self.kwargs = {}
+
+    def set_validation_batch(self, validation_batch_size=20):
+        """
+        Samples a batch of model traces and stores it as an object property.
+        """
+        self.validation_batch = [self._sample_from_joint()
+                                 for _ in range(validation_batch_size)]
+
+    def set_args(self, *args, **kwargs):
+        """
+        Arguments are stored and passed to both the model and guide during
+        training and inference.
+        """
+        self.args = args
+        self.kwargs = kwargs
+
+    def _traces(self, **kwargs):
+        """
+        Wrapping inherited _traces function to allow inference to be performed
+        with the arguments set in set_args. Intended to allow user to specify
+        keyword arguments containing observations when performing inference.
+        """
+        merged_kwargs = self.kwargs.copy()
+        for k, v in kwargs.items():
+            merged_kwargs[k] = v
+        return super(CSIS, self)._traces(*self.args, **merged_kwargs)
+
+    def step(self):
+        """
+        :returns: estimate of the loss
+        :rtype: float
+
+        Take a gradient step on the loss function. Any args or kwargs are
+        passed to the model and guide.
+        """
+        loss = self.loss(grads=True)
+        self.optim.step()
+        self.optim.zero_grad()
+
+        return torch_item(loss)
+
+    def loss(self, grads=False, batch=None):
```

Comment:
If you refactor this a little bit to look more like `infer.Trace_ELBO`, it'll be much easier to add support for the PyTorch JIT (which currently doesn't support calling `backward` within compiled functions).  In particular, you can rename `CSIS.loss` to `loss_and_grads`, define a `differentiable_loss_particle` method that computes `particle_loss` and call that inside the main loop in `loss_and_grads`.

Reply:
Sure, I've now refactored it to use `loss_and_grads` and `differentiable_loss_particle`. I'm not sure exactly how similar they should be though - e.g. `CSIS.loss_and_grads` still takes in an argument specifying whether to take gradients, unlike `Trace_ELBO.loss_and_grads`. Is this a problem?